spec/core/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -517,20 +566,140 @@ function chanCloseConfirm(
-      proofHeight,
-      proofInit,
-      channel.counterpartyPortIdentifier,
-      channel.counterpartyChannelIdentifier,
-      expected
-    ))
+                          channelIdentifier, counterpartyHops, channel.version}
+
+    if (connectionHops.length > 1) {
+      key = host.ChannelPath(counterparty.PortId, counterparty.ChannelId)
+
+      abortTransactionUnless(connection.verifyMultihopMembership(
+        connection,
+        proofHeight,
+        proofInit,
+        channel.connectionHops,
+        key
+        expected))
+    } else {
+      abortTransactionUnless(connection.verifyChannelState(
+        proofHeight,
+        proofInit,
+        channel.counterpartyPortIdentifier,
+        channel.counterpartyChannelIdentifier,
+        expected
+      ))
+    }
+
     channel.state = CLOSED
     provableStore.set(channelPath(portIdentifier, channelIdentifier), channel)
 }
 ```
 
+The `chanCloseFrozen` function is called by a relayer to force close a multi-hop channel if any client state in the
+channel path is frozen. A relayer should send proof of the frozen client state to each end of the channel with a
+proof of the frozen client state in the channel path starting from each channel end up until the first frozen client.
+The multi-hop proof for each channel end will be different and consist of a proof formed starting from each channel
+end up to the frozen client.
+
+The multi-hop proof starts with a chain with a frozen client for the misbehaving chain. However, the frozen client exists
+on the next blockchain in the channel path so the key/value proof is indexed to evaluate on the consensus state holding
+that client state. The client state path requires knowledge of the client id which can be determined from the
+connectionEnd on the misbehaving chain prior to the misbehavior submission.
+
+Once frozen, it is possible for a channel to be unfrozen (reactivated) via governance processes once the misbehavior in
+the channel path has been resolved. However, this process is out-of-protocol.
+
+```typescript
+function chanCloseFrozen(
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  proofConnection: MultihopProof,
+  proofClientState: MultihopProof,
+  proofHeight: Height) {
+    abortTransactionUnless(authenticateCapability(channelCapabilityPath(portIdentifier, channelIdentifier), capability))
+    channel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
+    abortTransactionUnless(channel !== null)
+    hopsLength = channel.connectionHops.length
+    abortTransactionUnless(hopsLength === 1)
+    abortTransactionUnless(channel.state !== CLOSED)
+    connection = provableStore.get(connectionPath(channel.connectionHops[0]))
+    abortTransactionUnless(connection !== null)
+    abortTransactionUnless(connection.state === OPEN)
+
+    // lookup connectionID for connectionEnd corresponding to misbehaving chain
+    let connectionIdx = proofConnection.ConsensusProofs.length + 1
```

Comment:
maybe?
```suggestion
    let connectionIdx = proofConnection.ConnectionProofs.length + 1
```
I assume that it is still the case that `ConnectionProofs.length == ConsensusProofs.length`.

Reply:
yes, this this is true and MUST be true...it bugs me that the current design allows them to differ :) I think ideally, they would be forced to be coupled by the message design so that they are guaranteed to have equal length and then no length checks would be required. I haven't gotten to these types of optimizations yet.